presto-manager-common/src/main/java/com/teradata/prestomanager/common/ApiRequester.java

```diff
@@ -30,37 +31,53 @@
 
 import static java.util.Objects.requireNonNull;
 
+@ThreadSafe
 public class ApiRequester
 {
-    private final UriBuilder uriBuilder;
+    private final Client client;
+    private final String uriTemplate;
     private final HttpMethod method;
     private final MultivaluedMap<String, Object> headers;
     private final String mediaType;
     @Nullable
-    private Entity entity;
+    private final Entity entity;
 
-    private Client client;
-
-    private ApiRequester(Client client, // TODO: Use AssistedInject to manage this
+    private ApiRequester(Client client,
             UriBuilder uriBuilder, HttpMethod method,
             Entity entity, MultivaluedMap<String, Object> headers,
             String mediaType)
     {
         this.client = requireNonNull(client);
-        this.uriBuilder = requireNonNull(uriBuilder);
+        this.uriTemplate = requireNonNull(uriBuilder).toTemplate();
         this.method = requireNonNull(method);
         this.headers = requireNonNull(headers);
         this.mediaType = requireNonNull(mediaType);
 
         this.entity = entity;
     }
 
+    /**
+     * Send this request to the given base URI.
+     * <p>
+     * The path should contain no components besides user-info, host, and port.
+     * <p>
+     * This method is thread-safe.
+     */
+    // TODO: The uri should be validated to not conflict with the uriTemplate
```

Comment:
It's less a matter of fixing this here than of determining where to fix it.

Should we validate URIs as they are discovered? If we validate here, should we only check for conflicting paths? Should we allow the input URI to add new parameters?